src/language/patterns.md

```diff
@@ -0,0 +1,334 @@
+
+For example, you can use object destructuring in a [for-in loop][for] to iterate-over
+and destructure the [`MapEntry`][] objects that a `<Map>.entries` call
+returns:
+
+```dart
+Map<String, int> hist = // ...
+
+for (var MapEntry(key: key, value: count) in hist.entries) {
+  print('$key occurred $count times');
+}
+```
+
+The object pattern checks that `hist.entries` has the named type `MapEntry`,
+and then recurses into the named field subpatterns `key` and `value`.
+It calls the `key` getter and `value` getter on the `MapEntry` in each iteration,
+and binds the results to local variables `key` and `count`, respectively.
+
+Binding the result of a getter call to a variable of the same name is a common
+use case, so object patterns can also infer the getter name from the
+[variable subpattern][variable]. This allows you to simplify the variable pattern
+from something redundant like `key: key` to just `:key`:
+
+```dart
+for (var MapEntry(:key, value: count) in hist.entries) { 
+  print('$key occurred $count times');
+}
+```
+
+### Algebraic datatypes 
+
+Object destructuring and switch cases are conducive to writing
+code in an [algebraic datatype][] style: 
+when you have a family of related types and an operation that needs
+specific behavior for each type.
+
+Instead of implementing the operation as an instance method for every respective type,
+you can keep the operations in a single function, switching over subtypes for
+variations in behavior:
+
+```dart
+abstract class Shape {
+  double calculateArea();
+}
+
+class Square implements Shape {
+  final double length;
+  Square(this.length);
+
+  double calculateArea() => length * length;
+}
+
+class Circle implements Shape {
+  final double radius;
+  Circle(this.radius);
+
+  double calculateArea() => math.pi * radius * radius;
+}
+
+double calculateArea(Shape shape) =>
+  switch (shape) {
+    Square(length: var l) => l * l,
+    Circle(radius: var r) => math.pi * r * r
+  };
+```
+
+### Validating incoming JSON
+
+[Map][] and [list][] patterns work well for destructuring key-value pairs in
+JSON data:
+
+```dart 
+var json = {'user': ['Lily', 13]};
+var {'user': [name, age]} = json;
+```
+
+The example above makes sense when you already know that the JSON
+data has the structure you expect.
+If the data comes from an external source, like over the network as is commonly
+the case, you can't be sure of it's structure and
+should validate it first. 
+
+Without patterns, validation is verbose:
+
+```dart
+if (json is Map<String, dynamic> &&
+    json.length == 1 && json.containsKey('user')) {
+  var user = json[user];
+  if (user is List<dynamic> && user.length == 2 &&
+      user[0] is String && user[1] is int) {
+    var name = user[2] as String;
+    var age = user[1] as int;
+    print('User $name is $age years old.');
+  }
+}
+```
+
+A single [case pattern](#switch-cases) in a switch achieves the same validation
+as above, but is much clearer, more declarative, and much shorter:
```

Comment:
This might be a good place to use `if-case` instead of a switch since there's only a single case.

Reply:
Changed it to:

> A single [case pattern][] (conducive to an [if-case][] statement)
> achieves the same validation as above, but is much clearer, more declarative,
> and much shorter:
> 
> ```dart
> if (json case {'user': [String name, int age]}) 
>   print('User $name is $age years old.');
> ```

The parenthetical about if-case, keep or toss?